diff --git a/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php b/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php
index e628605..e00addf 100644
--- a/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php
+++ b/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php
@@ -119,7 +119,13 @@ public function loadMultiple(array $ids = []) {
    * {@inheritdoc}
    */
   public function warmMultiple(array $items = []) {
+    $accountSwitcher = \Drupal::service('account_switcher');
+    $account = \Drupal::entityTypeManager()
+      ->getStorage('user')
+      ->load(1);
+    $accountSwitcher->switchTo($account);
     $normalizations = array_map([$this->entityToJsonapi, 'normalize'], $items);
+    $accountSwitcher->switchBack();
     count($normalizations);
   }

