diff --git a/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php b/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php
index e628605..0efa7b6 100644
--- a/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php
+++ b/modules/contrib/jsonapi_boost/src/Plugin/warmer/ResourceWarmer.php
@@ -5,6 +5,7 @@
 use Drupal\Core\Annotation\Translation;
 use Drupal\Core\Entity\EntityTypeManagerInterface;
 use Drupal\Core\Form\SubformStateInterface;
+use Drupal\Core\Session\UserSession;
 use Drupal\jsonapi\ResourceType\ResourceTypeRepositoryInterface;
 use Drupal\jsonapi_extras\Entity\JsonapiResourceConfig;
 use Drupal\jsonapi_extras\EntityToJsonApi;
@@ -119,8 +120,12 @@ public function loadMultiple(array $ids = []) {
    * {@inheritdoc}
    */
   public function warmMultiple(array $items = []) {
+    $current_user = \Drupal::service('current_user');
+    $anonymous = $current_user->getAccount();
+    $current_user->setAccount(new UserSession(['roles' => ['creative', 'editor']]));
     $normalizations = array_map([$this->entityToJsonapi, 'normalize'], $items);
     count($normalizations);
+    $current_user->setAccount($anonymous);
   }

   /**

