diff --git a/modules/contrib/image_style_warmer/src/BatchService.php b/modules/contrib/image_style_warmer/src/BatchService.php
index edf34cda..7d48180a 100644
--- a/modules/contrib/image_style_warmer/src/BatchService.php
+++ b/modules/contrib/image_style_warmer/src/BatchService.php
@@ -19,7 +19,7 @@ class BatchService {
    * @param object $context
    *   Context for operations.
    */
-  public function warmUpFileProcess($fid, $count, &$context) {
+  public static function warmUpFileProcess($fid, $count, &$context) {
     /** @var \Drupal\image_style_warmer\ImageStylesWarmerInterface $image_styles_warmer */
     $image_styles_warmer = \Drupal::service('image_style_warmer.warmer');
     $file = File::load($fid);
@@ -45,7 +45,7 @@ public function warmUpFileProcess($fid, $count, &$context) {
    * @param array $operations
    *   Array of operations.
    */
-  public function warmUpFileFinished($success, array $results, array $operations) {
+  public static function warmUpFileFinished($success, array $results, array $operations) {
     $messenger = \Drupal::messenger();
     if ($success) {
       $messenger->addMessage(t('@count files warmed up.', ['@count' => count($results)]));
