diff --git a/modules/contrib/image_style_warmer/src/ImageStylesWarmer.php b/modules/contrib/image_style_warmer/src/ImageStylesWarmer.php
index 5a25b3c4..19eea235 100644
--- a/modules/contrib/image_style_warmer/src/ImageStylesWarmer.php
+++ b/modules/contrib/image_style_warmer/src/ImageStylesWarmer.php
@@ -5,6 +5,7 @@
 use Drupal\Core\Config\ConfigFactoryInterface;
 use Drupal\Core\Entity\EntityTypeManagerInterface;
 use Drupal\Core\Image\ImageFactory;
+use Drupal\Core\Installer\InstallerKernel;
 use Drupal\Core\Queue\QueueFactory;
 use Drupal\file\FileInterface;

@@ -79,7 +80,12 @@ public function __construct(ConfigFactoryInterface $config_factory, EntityTypeMa
   public function warmUp(FileInterface $file) {
     $initialImageStyles = $this->config->get('initial_image_styles');
     if (!empty($initialImageStyles)) {
-      $this->doWarmUp($file, array_keys($initialImageStyles));
+      if (!InstallerKernel::installationAttempted()) {
+        $this->doWarmUp($file, array_keys($initialImageStyles));
+      }
+      else {
+        $this->addQueue($file, array_keys($initialImageStyles));
+      }
     }
     $queueImageStyles = $this->config->get('queue_image_styles');
     if (!empty($queueImageStyles)) {
