diff --git a/core/includes/install.core.inc b/core/includes/install.core.inc
index 6832eae707..8644b96e4a 100644
--- a/core/includes/install.core.inc
+++ b/core/includes/install.core.inc
@@ -873,6 +873,7 @@ function install_tasks($install_state) {
       'run' => $needs_translations ? INSTALL_TASK_RUN_IF_NOT_COMPLETED : INSTALL_TASK_SKIP,
     ],
     'install_finished' => [],
+    'install_invoke_finished_hooks' => [],
   ];

   // Allow the installation profile to modify the full list of tasks.
@@ -1884,6 +1885,16 @@ function install_finished(&$install_state) {
   \Drupal::state()->set('install_time', \Drupal::time()->getRequestTime());
 }

+/**
+ * Invokes hooks related to site installation being finished.
+ *
+ * @param $install_state
+ *   An array of information about the current installation state.
+ */
+function install_invoke_finished_hooks(&$install_state) {
+  \Drupal::moduleHandler()->invokeAll('site_install_finished', array_values($install_state));
+}
+
 /**
  * Implements callback_batch_operation().
  *
diff --git a/core/lib/Drupal/Core/Extension/module.api.php b/core/lib/Drupal/Core/Extension/module.api.php
index bab77e3a54..48ccd4eedd 100644
--- a/core/lib/Drupal/Core/Extension/module.api.php
+++ b/core/lib/Drupal/Core/Extension/module.api.php
@@ -513,6 +513,41 @@ function hook_install_tasks_alter(&$tasks, $install_state) {
   $tasks['install_configure_form']['function'] = 'my_profile_install_configure_form';
 }

+/**
+ * Responds to Drupal being installed.
+ *
+ * @param $install_state
+ *   An array of information about the current installation state.
+ *
+ * @see install_tasks()
+ */
+function hook_site_install_finished($install_state) {
+
+  // Example function body for setting state only once installation
+  // has finally completed.
+  if (!empty($install_state['parameters']['profile'])) {
+    $profile = $install_state['parameters']['profile'];
+    if ($profile === 'special_profile') {
+      \Drupal::state()->set('my_module_special_profile_support', TRUE);
+    }
+  }
+
+  // Example function body setting a status message only once installation
+  // has finally completed.
+  if (!empty($install_state['parameters']['profile'])) {
+    $profile = $install_state['parameters']['profile'];
+    if ($profile === 'demo_umami') {
+      $umami = 'https://www.drupal.org/docs/8/umami-drupal-8-demonstration-installation-profile';
+      \Drupal::messenger()->addStatus(t('Read more about <a href=":umami">Umami</a> online.',
+        [
+          ':umami' => $umami,
+        ])
+      );
+    }
+  }
+
+}
+
 /**
  * Perform a single update between minor versions.
  *
