diff --git a/src/Drush/Commands/WarmerCommands.php b/src/Drush/Commands/WarmerCommands.php
--- a/src/Drush/Commands/WarmerCommands.php
+++ b/src/Drush/Commands/WarmerCommands.php	(date 1734209465451)
@@ -104,6 +104,7 @@
       return $count;
     }, $warmers);
     $total = array_sum($count_list);
+    $items_limit = max($count_list);
     $this->logger()->success(
       dt('@total items enqueued for cache warming.', ['@total' => $total])
     );
@@ -115,7 +116,7 @@
     }
     $this->logger()->success(dt('Warming caches in @count batches from the "warmer" queue.', ['@count' => $batch_count]));

-    $this->processManager()->drush($this->siteAliasManager()->getSelf(), 'queue:run', ['warmer'], ['time-limit' => static::VERY_HIGH_NUMBER, 'items-limit' => $total]);
+    $this->processManager()->drush($this->siteAliasManager()->getSelf(), 'queue:run', ['warmer'], ['time-limit' => static::VERY_HIGH_NUMBER, 'items-limit' => $items_limit]);
   }

   /**
