diff --git a/src/Utility/FieldItem.php b/src/Utility/FieldItem.php
--- a/src/Utility/FieldItem.php
+++ b/src/Utility/FieldItem.php
@@ -59,7 +59,8 @@
     if (
       isset($element['#type'], $context['items']) &&
       $element['#type'] === 'managed_file' &&
-      $context['items'] instanceof FileFieldItemList
+      $context['items'] instanceof FileFieldItemList &&
+      $element['#entity_type'] !== 'consumer'
     ) {
       return $context['items'];
     }
