diff --git a/modules/contrib/filefield_paths/src/Utility/FieldItem.php b/modules/contrib/filefield_paths/src/Utility/FieldItem.php
--- a/modules/contrib/filefield_paths/src/Utility/FieldItem.php
+++ b/modules/contrib/filefield_paths/src/Utility/FieldItem.php
@@ -54,7 +54,7 @@
    *   Returns Field Item List instance. Null if widget type is not supported.
    */
   public static function getFromSupportedWidget(array $element, array $context): ?FileFieldItemList {
-    if(isset($element['#type']) && $element['#type'] === 'managed_file') {
+    if (isset($element['#type']) && $element['#type'] == 'managed_file' && $element['#entity_type'] != 'consumer') {
       return $context['items'] ?? NULL;
     }
     return NULL;
