diff --git a/modules/contrib/filefield_paths/filefield_paths.module b/web/modules/filefield_paths/filefield_paths.module
index dd207c5..dcfb440 100644
--- a/modules/contrib/filefield_paths/filefield_paths.module
+++ b/modules/contrib/filefield_paths/filefield_paths.module
@@ -211,7 +211,7 @@ function filefield_paths_form_field_config_edit_form_alter(array &$form, FormSta
 function filefield_paths_field_widget_form_alter(&$element, FormStateInterface $form_state, $context) {
   // Force all File (Field) Paths uploads to go to the temporary file system
   // prior to being processed.
-  if (isset($element['#type']) && $element['#type'] == 'managed_file') {
+  if (isset($element['#type']) && $element['#type'] == 'managed_file' && $element['#entity_type'] != 'consumer') {
     $settings = $context['items']->getFieldDefinition()
       ->getThirdPartySettings('filefield_paths');
     if (isset($settings['enabled']) && $settings['enabled']) {

