diff --git a/modules/contrib/jsonapi_cross_bundles/src/ResourceType/CrossBundlesResourceType.php b/modules/contrib/jsonapi_cross_bundles/src/ResourceType/CrossBundlesResourceType.php
index de45180..83cd8a9 100644
--- a/modules/contrib/jsonapi_cross_bundles/src/ResourceType/CrossBundlesResourceType.php
+++ b/modules/contrib/jsonapi_cross_bundles/src/ResourceType/CrossBundlesResourceType.php
@@ -54,7 +54,7 @@ public function getBundleResourceTypes() {
    * {@inheritdoc}
    */
   public function getPath() {
-    return $this->entityTypeId;
+    return $this->entityTypeId . 's';
   }

   /**
