diff --git a/src/Normalizer/JsonBlueprintDenormalizer.php b/src/Normalizer/JsonBlueprintDenormalizer.php
index 6611514adc8089c5c654e35d028a2f5b87359a1d..c1cc038829efb9be56a0a5fc56717cd3f8c0cae3 100644
--- a/src/Normalizer/JsonBlueprintDenormalizer.php
+++ b/src/Normalizer/JsonBlueprintDenormalizer.php
@@ -53,7 +53,7 @@ class JsonBlueprintDenormalizer implements DenormalizerInterface, SerializerAwar
   /**
    * Sets the validator service if available.
    */
-  public function setValidator(Validator $validator = NULL) {
+  public function setValidator(?Validator $validator = NULL) {
     if ($validator) {
       $this->validator = $validator;
     }

