diff --git a/modules/contrib/multivalue_form_element/src/Element/MultiValue.php b/modules/contrib/multivalue_form_element/src/Element/MultiValue.php
index 7d4246b..e57ba5a 100644
--- a/modules/contrib/multivalue_form_element/src/Element/MultiValue.php
+++ b/modules/contrib/multivalue_form_element/src/Element/MultiValue.php
@@ -301,6 +301,18 @@ public static function validateMultiValue(array &$element, FormStateInterface $f
    */
   public static function valueCallback(&$element, $input, FormStateInterface $form_state) {
     if ($input !== FALSE) {
+      $forms_to_modify = [
+        'question_radios_edit_form',
+        'question_checkboxes_edit_form',
+        'paragraph_questionnaire_edit_form',
+      ];
+      if (in_array($form_state->getFormObject()->getFormId(), $forms_to_modify)) {
+        foreach ($input as $key => $in) {
+          if (empty($in['correct']) && !empty($in['option'])) {
+            $input[$key]['correct'] = 0;
+          }
+        }
+      }
       return $input;
     }

