diff --git a/src/JsonapiParse.php b/src/JsonapiParse.php
--- a/src/JsonapiParse.php
+++ b/src/JsonapiParse.php
@@ -47,7 +47,9 @@
    */
   #[\Override]
   public function parse(Response $response): Response {
-    $this->parseJsonContent($response);
+    $content = Json::decode($response->getContent() ?: '') ?? [];
+    $content = $this->parseJsonContent($content);
+    $response->setContent(Json::encode($content));
     return $response;
   }
 
@@ -177,10 +179,10 @@
    * @param string $key
    *   Relationship key.
    *
-   * @return array
+   * @return array|null
    *   Result.
    */
-  protected function resolveRelationshipData(mixed $links, string $key): array {
+  protected function resolveRelationshipData(mixed $links, string $key): ?array {
     if (empty($links['data'])) {
       return $links;
     }
@@ -236,24 +238,18 @@
   /**
    * Integrates includes into the content of a Response.
    *
-   * @param \Symfony\Component\HttpFoundation\Response|string|array $response
-   *   A Response object or string/array with a response content.
+   * @param array $content
+   *   An array with JSON:API response content.
    *
-   * @return \Symfony\Component\HttpFoundation\Response|array
-   *   Returns an array with the response content, if the input is string or
-   *   array, or void if input is a Response.
+   * @return array
+   *   The parsed array with the JSON:API response content.
    */
-  protected function parseJsonContent(Response $response): Response|array {
-    $content = $response->getContent();
-    if (is_string($content)) {
-      $content = Json::decode($content);
-    }
-
+  protected function parseJsonContent(array $content): array {
     if (NestedArray::getValue($content, ['jsonapi', 'parsed'])) {
-      return $response;
+      return $content;
     }
     if (isset($content['errors']) || empty($content['data'])) {
-      return $response;
+      return $content;
     }
     $this->included = $this->groupIncludes($content);
     $include_parameter = $this->requestStack->getCurrentRequest()->query->get('include');
@@ -272,8 +268,7 @@
     }
     $content['jsonapi']['parsed'] = TRUE;
     $content['data'] = $data;
-    $response->setContent(Json::encode($content));
-    return $response;
+    return $content;
   }
 
 }
diff --git a/src/JsonapiParseInterface.php b/src/JsonapiParseInterface.php
--- a/src/JsonapiParseInterface.php
+++ b/src/JsonapiParseInterface.php
@@ -12,13 +12,13 @@
 interface JsonapiParseInterface {
 
   /**
-   * Parses the JSON:API Response with integrating includes inside the fields.
+   * Parses the JSON:API response with integrating includes inside the fields.
    *
    * @param \Symfony\Component\HttpFoundation\Response $response
-   *   A Response object with JSON:API response.
+   *   A response object with JSON:API content.
    *
    * @return \Symfony\Component\HttpFoundation\Response
-   *   Returns the Response object.
+   *   The altered response.
    */
   public function parse(Response $response): Response;
 
