diff --git a/modules/contrib/jsonapi_include/src/JsonapiParse.php b/modules/contrib/jsonapi_include/src/JsonapiParse.php
index 9cc4e09..78a4414 100644
--- a/modules/contrib/jsonapi_include/src/JsonapiParse.php
+++ b/modules/contrib/jsonapi_include/src/JsonapiParse.php
@@ -250,7 +250,7 @@ protected function parseJsonContent($response) {
       return $json;
     }
     $this->included = $this->groupIncludes($json);
-    $include_parameter = \Drupal::request()->query->get('include');
+    $include_parameter = \Drupal::request()->query->get('include') ?? '';
     $this->allowed = array_map('trim', explode(',', $include_parameter));
     $data = [];
     if (!$this->isAssoc($json['data'])) {
