diff --git a/core/modules/jsonapi/jsonapi.services.yml b/core/modules/jsonapi/jsonapi.services.yml
index 56e585f..ef90598 100644
--- a/core/modules/jsonapi/jsonapi.services.yml
+++ b/core/modules/jsonapi/jsonapi.services.yml
@@ -1,5 +1,6 @@
 parameters:
   jsonapi.base_path: /jsonapi
+  jsonapi.include_drupal_ids: false

 services:
   jsonapi.serializer:
diff --git a/core/modules/jsonapi/src/JsonApiResource/ResourceIdentifier.php b/core/modules/jsonapi/src/JsonApiResource/ResourceIdentifier.php
index ca8b6e4..4eb59fe 100644
--- a/core/modules/jsonapi/src/JsonApiResource/ResourceIdentifier.php
+++ b/core/modules/jsonapi/src/JsonApiResource/ResourceIdentifier.php
@@ -297,7 +297,9 @@ public static function toResourceIdentifier(EntityReferenceItem $item, $arity =
     if (!is_null($arity)) {
       $meta[static::ARITY_KEY] = $arity;
     }
-    $meta["drupal_internal__$main_property_name"] = $properties[$main_property_name];
+    if (\Drupal::getContainer()->getParameter('jsonapi.include_drupal_ids')) {
+      $meta["drupal_internal__$main_property_name"] = $properties[$main_property_name];
+    }
     return new static($resource_type, $target->uuid(), $meta);
   }
