diff --git a/core/includes/install.core.inc b/core/includes/install.core.inc
index c8103369b6..06aab93fed 100644
--- a/core/includes/install.core.inc
+++ b/core/includes/install.core.inc
@@ -878,6 +878,7 @@ function install_tasks($install_state) {
       'run' => $needs_translations ? INSTALL_TASK_RUN_IF_NOT_COMPLETED : INSTALL_TASK_SKIP,
     ],
     'install_finished' => [],
+    'install_invoke_finished_hooks' => [],
   ];

   // Allow the installation profile to modify the full list of tasks.
@@ -1871,6 +1872,16 @@ function install_finished(&$install_state) {
   \Drupal::messenger()->addStatus($success_message);
 }

+/**
+ * Invokes hooks related to site installation being finished.
+ *
+ * @param $install_state
+ *   An array of information about the current installation state.
+ */
+function install_invoke_finished_hooks(&$install_state) {
+  \Drupal::moduleHandler()->invokeAll('site_install_finished', [$install_state]);
+}
+
 /**
  * Implements callback_batch_operation().
  *
diff --git a/core/lib/Drupal/Core/Extension/module.api.php b/core/lib/Drupal/Core/Extension/module.api.php
index e2ffe1320b..11561da340 100644
--- a/core/lib/Drupal/Core/Extension/module.api.php
+++ b/core/lib/Drupal/Core/Extension/module.api.php
@@ -502,6 +502,30 @@ function hook_install_tasks_alter(&$tasks, $install_state) {
   $tasks['install_configure_form']['function'] = 'myprofile_install_configure_form';
 }

+/**
+ * Perform necessary actions after Drupal being installed.
+ *
+ * It is recommended that you implement this hook if you want to perform some
+ * actions after site is installed.
+ *
+ * This hook is invoked at the end of installation profile.
+ *
+ * @param $install_state
+ *   An array of information about the current installation state.
+ *
+ * @see install_tasks()
+ */
+function hook_site_install_finished($install_state) {
+  // Example function body for setting state only once installation
+  // has finally completed.
+  if (!empty($install_state['parameters']['profile'])) {
+    $profile = $install_state['parameters']['profile'];
+    if ($profile === 'special_profile') {
+      \Drupal::state()->set('my_module_special_profile_support', TRUE);
+    }
+  }
+}
+
 /**
  * Perform a single update between minor versions.
  *

