diff --git a/src/Entities/AccessTokenEntity.php b/src/Entities/AccessTokenEntity.php
index 9c077fe..3e2ce84 100644
--- a/src/Entities/AccessTokenEntity.php
+++ b/src/Entities/AccessTokenEntity.php
@@ -2,10 +2,10 @@
 
 namespace Drupal\simple_oauth\Entities;
 
-use Lcobucci\JWT\Builder;
-use Lcobucci\JWT\Signer\Key;
+use DateTimeImmutable;
+use Lcobucci\JWT\Configuration;
+use Lcobucci\JWT\Signer\Key\InMemory;
 use Lcobucci\JWT\Signer\Rsa\Sha256;
-use League\OAuth2\Server\CryptKey;
 use League\OAuth2\Server\Entities\AccessTokenEntityInterface;
 use League\OAuth2\Server\Entities\Traits\AccessTokenTrait;
 use League\OAuth2\Server\Entities\Traits\EntityTrait;
@@ -18,31 +18,37 @@ class AccessTokenEntity implements AccessTokenEntityInterface {
   /**
    * {@inheritdoc}
    */
-  public function convertToJWT(CryptKey $privateKey) {
+  public function convertToJWT() {
     $private_claims = [];
-    \Drupal::moduleHandler()->alter('simple_oauth_private_claims', $private_claims, $this);
+    \Drupal::moduleHandler()
+      ->alter('simple_oauth_private_claims', $private_claims, $this);
     if (!is_array($private_claims)) {
       $message = 'An implementation of hook_simple_oauth_private_claims_alter ';
       $message .= 'returns an invalid $private_claims value. $private_claims ';
       $message .= 'must be an array.';
       throw new \InvalidArgumentException($message);
     }
-    $builder = (new Builder())
-      ->setAudience($this->getClient()->getIdentifier())
-      ->setId($this->getIdentifier(), TRUE)
-      ->setIssuedAt(time())
-      ->setNotBefore(time())
-      ->setExpiration($this->getExpiryDateTime()->getTimestamp())
-      ->setSubject($this->getUserIdentifier())
-      ->set('scopes', $this->getScopes());
+
+    $now = new DateTimeImmutable('@' . \Drupal::time()->getCurrentTime());
+    $key_path = $this->privateKey->getKeyPath();
+    $key = InMemory::file($key_path);
+    $config = Configuration::forSymmetricSigner(new Sha256(), $key);
+
+    $builder = $config->builder()
+      ->permittedFor($this->getClient()->getIdentifier())
+      ->identifiedBy($this->getIdentifier(), TRUE)
+      ->issuedAt($now)
+      ->canOnlyBeUsedAfter($now)
+      ->expiresAt($this->getExpiryDateTime())
+      ->relatedTo($this->getUserIdentifier())
+      ->withClaim('scopes', $this->getScopes());
 
     foreach ($private_claims as $claim_name => $value) {
-      $builder->set($claim_name, $value);
+      $builder->withClaim($claim_name, $value);
     }
 
-    $key = new Key($privateKey->getKeyPath(), $privateKey->getPassPhrase());
-    $token = $builder->sign(new Sha256(), $key)->getToken();
-    return $token;
+    return $builder->getToken($config->signer(), $config->signingKey());
   }
 
 }
+
