diff --git a/modules/contrib/simple_oauth/src/Authentication/Provider/SimpleOauthAuthenticationProvider.php b/modules/contrib/simple_oauth/src/Authentication/Provider/SimpleOauthAuthenticationProvider.php
index b55b3ce..976e741 100644
--- a/modules/contrib/simple_oauth/src/Authentication/Provider/SimpleOauthAuthenticationProvider.php
+++ b/modules/contrib/simple_oauth/src/Authentication/Provider/SimpleOauthAuthenticationProvider.php
@@ -75,6 +75,13 @@ public function authenticate(Request $request) {
       $auth_request = $this->resourceServer->validateAuthenticatedRequest($request);
     }
     catch (OAuthServerException $exception) {
+
+      // Silence token revoked messages.
+      if (str_contains($exception->getMessage(), 'The refresh token is invalid') ||
+        str_contains($exception->getMessage(), 'The resource owner or authorization server denied the request')) {
+        return NULL;
+      }
+
       // Procedural code here is hard to avoid.
       watchdog_exception('simple_oauth', $exception);
